api/v1/verticadb_webhook.go

```diff
@@ -1701,6 +1728,211 @@ func (v *VerticaDB) checkSandboxPrimary(allErrs field.ErrorList, oldObj *Vertica
+			}
+		}
+	}
+	return allErrs
+}
+
+// checkAnnotatedSubclustersInShutdownSandbox ensures: when a sandbox is to be shutdown, if a subcluster in it has annotation
+// vertica.com/shutdown-driven-by-sandbox set to true, the subcluster's shutdown field should not be updated
+func (v *VerticaDB) checkAnnotatedSubclustersInShutdownSandbox(old runtime.Object, allErrs field.ErrorList) field.ErrorList {
+	oldObj := old.(*VerticaDB)
+	persistSandboxes := v.findPersistSandboxes(oldObj)
+	newSandboxMap := v.GenSandboxMap()
+	oldSandboxMap := oldObj.GenSandboxMap()
+	newSubclusterIndexMap := v.GenSubclusterIndexMap()
+	newSubclusterMap := v.GenSubclusterMap()
+	oldSubclusterMap := oldObj.GenSubclusterMap()
+	for sandboxName := range persistSandboxes {
+		newSandbox := newSandboxMap[sandboxName]
+		oldSandbox := oldSandboxMap[sandboxName]
+		if newSandbox.Shutdown {
+			for _, subclusterName := range oldSandbox.Subclusters {
+				oldSubcluster := oldSubclusterMap[subclusterName.Name]
+				if vmeta.GetShutdownDrivenBySandbox(oldSubcluster.Annotations) {
+					newSubcluster, oldSclusterPersist := newSubclusterMap[subclusterName.Name]
+					if oldSclusterPersist && oldSubcluster.Shutdown != newSubcluster.Shutdown {
+						index := newSubclusterIndexMap[subclusterName.Name]
+						p := field.NewPath("spec").Child("subclusters")
+						err := field.Invalid(p.Index(index).Child("shutdown"),
+							newSubcluster.Shutdown,
+							fmt.Sprintf("cannot change Shutdown field for subcluster %q whose shutdown is driven by sandbox %q",
+								subclusterName.Name, oldSandbox.Name))
+						allErrs = append(allErrs, err)
+					}
+				}
+			}
+		}
+	}
+	return allErrs
+}
+
+// checkShutdownSandboxImage ensures a sandbox's image is not updated after the sandbox or any of its subclusters are shut down
+func (v *VerticaDB) checkShutdownSandboxImage(old runtime.Object, allErrs field.ErrorList) field.ErrorList {
+	oldObj := old.(*VerticaDB)
+	persistSandboxes := v.findPersistSandboxes(oldObj)
+	oldSandboxMap := oldObj.GenSandboxMap()
+	newSandboxMap := v.GenSandboxMap()
+	newSandboxIndexMap := v.GenSandboxIndexMap()
+	newSubclusterMap := v.GenSubclusterMap()
+	statusSClusterIndexMap := v.GenStatusSClusterIndexMap()
+	for sandboxName := range persistSandboxes {
+		oldSandbox := oldSandboxMap[sandboxName]
+		newSandbox := newSandboxMap[sandboxName]
+		if oldSandbox.Image != newSandbox.Image {
+			newSandboxIndex := newSandboxIndexMap[sandboxName]
+			errMsgs := v.checkSboxForShutdown(newSandbox, newSubclusterMap, statusSClusterIndexMap)
+			if len(errMsgs) != 0 {
+				p := field.NewPath("spec").Child("sandboxes").Index(newSandboxIndex)
+				err := field.Invalid(p.Child("image"),
+					newSandbox.Image,
+					fmt.Sprintf("cannot change the image for sandbox %q because %q",
+						sandboxName, strings.Join(errMsgs, ",")))
+				allErrs = append(allErrs, err)
+			}
+		}
+	}
+	return allErrs
+}
+
+// checkTerminatingSandboxes ensure the sandbox to terminate is not shut down
```

Comment:
This comment doesn't match the function implementation. I don't think we need to check if all subclusters inside the sandbox are terminating. If a "shutdown" sandbox is removed, we should error out.

Reply:
I will change the code to adopt the new logic. Could you update the rule in the JIRA as well?